pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -43,7 +43,144 @@
 # after mcp imports so any import error maps to this file, not _mcp.py
 from . import _mcp, _utils, exceptions, messages, models
 
-__all__ = 'MCPServer', 'MCPServerStdio', 'MCPServerHTTP', 'MCPServerSSE', 'MCPServerStreamableHTTP', 'load_mcp_servers'
+__all__ = (
+    'MCPServer',
+    'MCPServerStdio',
+    'MCPServerHTTP',
+    'MCPServerSSE',
+    'MCPServerStreamableHTTP',
+    'load_mcp_servers',
+    'MCPError',
+    'Resource',
+    'ResourceAnnotations',
+    'ResourceTemplate',
+    'ServerCapabilities',
+)
+
+
+class MCPError(RuntimeError):
+    """Raised when an MCP server returns an error response.
+
+    This exception wraps error responses from MCP servers, following the ErrorData schema
+    from the MCP specification.
+    """
+
+    message: str
+    """The error message."""
+
+    code: int
+    """The error code returned by the server."""
+
+    data: dict[str, Any] | None
+    """Additional information about the error, if provided by the server."""
+
+    def __init__(self, message: str, code: int, data: dict[str, Any] | None = None):
+        self.message = message
+        self.code = code
+        self.data = data
+        super().__init__(message)
+
+    @classmethod
+    def from_mcp_sdk_error(cls, error: mcp_exceptions.McpError) -> MCPError:
+        """Create an MCPError from an MCP SDK McpError.
+
+        Args:
+            error: An McpError from the MCP SDK.
+
+        Returns:
+            A new MCPError instance with the error data.
+        """
+        # Extract error data from the McpError.error attribute
+        error_data = error.error
+        return cls(message=error_data.message, code=error_data.code, data=error_data.data)
```

Comment:
Why do we need to do this?

Reply:
I mostly added the `... from e` to preserve the originating stack trace. You are right though, aside from the stack trace, our native exceptions literally replicate the data in the original exception 1:1.

If we don't think the originating stack trace is useful, I can remove it.

I still think it's useful to have the actual error details in the exception being thrown because it's easier for the user to access for "normal" error handling, eg:
```python
try:
    ...
except MCPError as e:
    print(f'Something terrible happened reading your resource: {e.message} (err: {e.code})')
```

Lemme know if I'm missing something!